File: java-app/src/main/java/com/k8smanager/service/KubernetesService.java
package com.k8smanager.service;

import io.kubernetes.client.openapi.ApiClient;
import io.kubernetes.client.openapi.ApiException;
import io.kubernetes.client.openapi.Configuration;
import io.kubernetes.client.openapi.apis.AppsV1Api;
import io.kubernetes.client.openapi.apis.CoreV1Api;
import io.kubernetes.client.openapi.models.V1Pod;
import io.kubernetes.client.openapi.models.V1PodList;
import io.kubernetes.client.openapi.models.V1Deployment;
import io.kubernetes.client.openapi.models.V1DeploymentList;
import io.kubernetes.client.util.Config;

import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.ArrayList;

public class KubernetesService {
    private final CoreV1Api coreApi;
    private final AppsV1Api appsApi;
    
    public KubernetesService() throws Exception {
        ApiClient client = Config.defaultClient();
        Configuration.setDefaultApiClient(client);
        this.coreApi = new CoreV1Api();
        this.appsApi = new AppsV1Api();
    }
    
    public Map<String, Object> getClusterInfo() throws ApiException {
        Map<String, Object> clusterInfo = new HashMap<>();
        
        // Get all pods - fixed method signature with 11 parameters
        V1PodList podList = coreApi.listPodForAllNamespaces(
            null, null, null, null, null, null, null, null, null, null, null);
        
        // Get all deployments - fixed method signature with 11 parameters  
        V1DeploymentList deploymentList = appsApi.listDeploymentForAllNamespaces(
            null, null, null, null, null, null, null, null, null, null, null);
        
        // Convert to simple data structures for JSON serialization
        List<Map<String, Object>> pods = new ArrayList<>();
        for (V1Pod pod : podList.getItems()) {
            Map<String, Object> podInfo = new HashMap<>();
            podInfo.put("name", pod.getMetadata().getName());
            podInfo.put("namespace", pod.getMetadata().getNamespace());
            podInfo.put("status", pod.getStatus().getPhase());
            // Convert timestamp to string to avoid serialization issues
            if (pod.getMetadata().getCreationTimestamp() != null) {
                podInfo.put("creationTimestamp", pod.getMetadata().getCreationTimestamp().toString());
            }
            pods.add(podInfo);
        }
        
        List<Map<String, Object>> deployments = new ArrayList<>();
        for (V1Deployment deployment : deploymentList.getItems()) {
            Map<String, Object> deploymentInfo = new HashMap<>();
            deploymentInfo.put("name", deployment.getMetadata().getName());
            deploymentInfo.put("namespace", deployment.getMetadata().getNamespace());
            deploymentInfo.put("replicas", deployment.getSpec().getReplicas());
            deploymentInfo.put("readyReplicas", deployment.getStatus().getReadyReplicas());
            // Convert timestamp to string
            if (deployment.getMetadata().getCreationTimestamp() != null) {
                deploymentInfo.put("creationTimestamp", deployment.getMetadata().getCreationTimestamp().toString());
            }
            deployments.add(deploymentInfo);
        }
            
        clusterInfo.put("pods", pods);
        clusterInfo.put("deployments", deployments);
        clusterInfo.put("podCount", pods.size());
        clusterInfo.put("deploymentCount", deployments.size());
        
        return clusterInfo;
    }
}

File: python-k8s-manager/services/kubernetes_service.py
from kubernetes import client, config
from kubernetes.client.rest import ApiException
from typing import List, Optional
import time
from models.cluster_data import ClusterData, PodInfo, DeploymentInfo

class KubernetesService:
    """Service for interacting with Kubernetes API"""
    
    def __init__(self):
        try:
            # Load in-cluster config
            config.load_incluster_config()
            print("Loaded in-cluster Kubernetes config")
        except:
            try:
                # Fallback to local config for development
                config.load_kube_config()
                print("Loaded local Kubernetes config")
            except Exception as e:
                print(f"Failed to load Kubernetes config: {e}")
                raise
        
        self.core_v1 = client.CoreV1Api()
        self.apps_v1 = client.AppsV1Api()
    
    def fetch_cluster_data(self) -> ClusterData:
        """Fetch cluster data from Kubernetes API"""
        try:
            start_time = time.time()
            print("Fetching cluster data from Kubernetes API...")
            
            # Fetch pods
            pods = self._fetch_pods()
            
            # Fetch deployments
            deployments = self._fetch_deployments()
            
            duration = time.time() - start_time
            print(f"Cluster data fetch completed in {duration:.2f}s - {len(pods)} pods, {len(deployments)} deployments")
            
            return ClusterData(
                pods=pods,
                deployments=deployments,
                pod_count=len(pods),
                deployment_count=len(deployments),
                fetch_timestamp=time.time()
            )
            
        except ApiException as e:
            print(f"Kubernetes API error: {e}")
            raise
        except Exception as e:
            print(f"Error fetching cluster data: {e}")
            raise
    
    def _fetch_pods(self) -> List[PodInfo]:
        """Fetch all pods from all namespaces"""
        try:
            pod_list = self.core_v1.list_pod_for_all_namespaces()
            pods = []
            
            for pod in pod_list.items:
                pod_info = PodInfo(
                    name=pod.metadata.name,
                    namespace=pod.metadata.namespace,
                    status=pod.status.phase or "Unknown",
                    creation_timestamp=pod.metadata.creation_timestamp.isoformat() if pod.metadata.creation_timestamp else None
                )
                pods.append(pod_info)
            
            return pods
            
        except ApiException as e:
            print(f"Error fetching pods: {e}")
            raise
    
    def _fetch_deployments(self) -> List[DeploymentInfo]:
        """Fetch all deployments from all namespaces"""
        try:
            deployment_list = self.apps_v1.list_deployment_for_all_namespaces()
            deployments = []
            
            for deployment in deployment_list.items:
                deployment_info = DeploymentInfo(
                    name=deployment.metadata.name,
                    namespace=deployment.metadata.namespace,
                    replicas=deployment.spec.replicas,
                    ready_replicas=deployment.status.ready_replicas,
                    creation_timestamp=deployment.metadata.creation_timestamp.isoformat() if deployment.metadata.creation_timestamp else None
                )
                deployments.append(deployment_info)
            
            return deployments
            
        except ApiException as e:
            print(f"Error fetching deployments: {e}")
            raise
